docs/tutorials/PYLS-guide.md

```diff
@@ -0,0 +1,86 @@
+
+This guide will show you how to configure `lsp-mode` for Python development.
+There are several language servers available for python:
+  * [pyls](https://github.com/palantir/python-language-server) by Palantir Technologies (this tutorial focuses on this ls)
+  * [python-ms](https://github.com/Microsoft/python-language-server) by Microsoft
+  * [jedi](https://github.com/pappasam/jedi-language-server) by pappasam
+  * [pyright](https://github.com/microsoft/pyright) by Microsoft
+
+## Features of `pyls` as a language server
+
+* Code completion
+* Linting and syntax checking
+* Signature help
+* Code navigation (references/definitions/etc..)
+* Info on hover
+* Document Formatting
+
+# Installing the language server
+
+There are two ways to install `pyls`:
+1. Make a global install
+2. Installation per-project
+
+Per project installation is a tad bit faster but almost unnoticeable, so if you find global installation
+a bit too slow and you have exhausted your [optimization](#optimizations) options, try a per-project approach.
+
+## Global install
+
+1. Install globally with `pip install --user python-language-server[all]`
+2. If your project uses virtualenv in any form (`virtualenv`, `virtualenvwrapper`, `poetry`, etc...) you need to specify a few project specific configurations
+    * Go to the root of your project
+    * Create a `.dir-locals.el`  file. Also, preferably, add it to your `.gitignore`
+    * place this code there:
+    ```elisp
+    ((python-mode .
+                  ((lsp-pyls-plugins-jedi-environment . "*ABSOLUTE PATH TO YOUR VIRTUALENV*"))))
+    ```
+    if you use virtualenvwrapper, you can switch it a bit
+    ```elisp
+    ((python-mode .
+                  ((lsp-pyls-plugins-jedi-environment . (concat (file-name-as-directory (getenv "WORKON_HOME")) "*venv name*")))))
+    ```
+
+## Installation per-project
+
+Per-project installation allows you to have different versions of `pyls` for your different projects.
+It requires your project to have some sort of virtualenv and a few more steps.
+
+1. Install `pyls` inside your virtualenv
+```bash
+source $PATH_TO_YOUR_VENV/bin/activate
+pip install python-language-server[all]
+```
+2. You need to make emacs know about virtualenv, and you need to do it **before** a call to `(lsp)`
+    * If you use Spacemacs \
+    You are in luck. Just create a file `.venv` with a path to your virtualenv in the root of your project.
+    * If you use Doom Emacs \
+    TBA (try the vanilla emacs approach, as well as check this [issue](https://github.com/hlissner/doom-emacs/issues/1666) in doom repo)
+    * If you use vanilla Emacs \
+    Check how spacemacs [does it](https://github.com/syl20bnr/spacemacs/blob/a9fc0d5fcc46aaef95aef99de3097fd79f0e3d26/layers/%2Blang/python/funcs.el#L225) with the [pyvenv](https://github.com/jorgenschaefer/pyvenv) emacs package \
+    It should look something like this (ignoring other options except `:hook`):
+    ```elisp
+    (use-package lsp-mode
```

Comment:
Can we have some self-contained configuration like the one in cpp guide or this is not applicable for python? I wish there is a script that "just-works" and that can be used as a base.

Reply:
I'll look into it more. But python is a bit too dynamic for that with different virtualenv tools emerging as fast as new js frameworks. 
If you are about the per-project approach being a bit too cumbersome to set up - it is. So we can specify the global installation as a recommended way, while leaving the per-project approach on the mercy of the user.
With global install we can try to make an educated guess for the virtualenv location, kinda like spacemacs does, and set the `pyls.jedi.environment` ourselves in `(lsp)` call somewhere during configuration of the client. 

The only issue I see with this is if we guess wrong, which we easily can, how do we allow the user to specify a proper virtualenv? Because, afaiu, dir-locals won't work, we'd just overwrite them...

Reply:
Dir locals work afaik.

I want the tutorial to be in the form "if you follow each of the steps it will just work" with all of the steps. E. g. you may use the recommended tool for creating a project skeleton(or the one you like) then describe a flow. I wish venvs usecase to be covered because this question often pops up but no need to have a detailed description, just one way to make it work. E. g. like this - https://code.visualstudio.com/docs/nodejs/reactjs-tutorial even if you have zero JS experience you will be able to follow it.

Reply:
Okay, I will only leave a global install as it is more straightforward